packages/core/src/codecatalyst/devEnv.ts

```diff
@@ -11,6 +11,122 @@ import { Timeout } from '../shared/utilities/timeoutUtils'
 import { showMessageWithCancel } from '../shared/utilities/messages'
 import { isCloud9 } from '../shared/extensionUtilities'
 import { getLogger } from '../shared/logger'
+import { CodeCatalystAuthenticationProvider } from './auth'
+import { getThisDevEnv } from './model'
+import { isInDevEnv } from '../shared/vscode/env'
+import { shared } from '../shared/utilities/functionUtils'
+
+/** This class exists due to complexities with starting the Dev Env Activity Hearbeat mechanism */
+export class DevEnvActivityHeartbeatStarter {
+    /**
+     * Trys to start the Dev Env Activity Heartbeat mechanism.
+     * This mechanism keeps the Dev Env from timing out.
+     */
+    public static registerDevEnvActivityHeartbeatStarter(authProvider: CodeCatalystAuthenticationProvider) {
+        if (!isInDevEnv()) {
+            getLogger().debug('codecatalyst: not in a devenv, not registering DevEnvActivityHeartbeatStarter')
+            return
+        }
+
+        DevEnvActivityHeartbeatStarter.authProvider = authProvider
+        void DevEnvActivityHeartbeatStarter.instance.tryStartDevEnvActivityHeartbeatWithRetry()
+    }
+
+    /** If true, the Activity Heartbeat is running */
+    private didStartDevEnvActivityHeartbeat: boolean = false
+    private startHeartbeatInterval: NodeJS.Timer | undefined
+
+    protected constructor(private readonly authProvider: CodeCatalystAuthenticationProvider) {}
+
+    private static _instance: DevEnvActivityHeartbeatStarter | undefined = undefined
+    private static get instance(): DevEnvActivityHeartbeatStarter {
+        return (DevEnvActivityHeartbeatStarter._instance ??= new DevEnvActivityHeartbeatStarter(
+            DevEnvActivityHeartbeatStarter.getAuthProvider()
+        ))
+    }
+
+    /** Keeps executing {@link DevEnvActivityHeartbeatStarter.tryStartDevEnvActivityHeartbeat} on an interval until it succeeds*/
+    private async tryStartDevEnvActivityHeartbeatWithRetry() {
+        // try immediately since interval below delays
+        await this.tryStartDevEnvActivityHeartbeat()
+
+        this.startHeartbeatInterval ??= globals.clock.setInterval(async () => {
+            await this.tryStartDevEnvActivityHeartbeat()
+
+            if (this.didStartDevEnvActivityHeartbeat) {
+                // on success stop the interval
+                this.startHeartbeatInterval?.unref()
+                this.startHeartbeatInterval = undefined
+            }
+        }, 30_000)
+    }
+
+    /** Trys to start the Activity Heartbeat mechanism */
+    private tryStartDevEnvActivityHeartbeat = shared(this._tryStartDevEnvActivityHeartbeat.bind(this)) // prevents race conditions
+    private async _tryStartDevEnvActivityHeartbeat() {
```

Comment:
probably can avoid the extra function by using arrow function (which binds to `this`)


```suggestion
    private tryStartDevEnvActivityHeartbeat = shared(async () => {
```

Reply:
nice